support/cas-server-support-saml-idp-web/src/main/java/org/apereo/cas/support/saml/web/idp/profile/builders/response/BaseSamlProfileSamlResponseBuilder.java

```diff
@@ -97,10 +97,12 @@ protected T encodeFinalResponse(final HttpServletRequest request,
 
         if (encodeResponse) {
             val context = new JEEContext(request, response);
-            val relayState = samlResponseBuilderConfigurationContext.getSessionStore()
+            val relayStateRequest = request.getParameter(SamlProtocolConstants.PARAMETER_SAML_RELAY_STATE);
```

Comment:
Fix generally looks OK. However, this part is not necessary. Rather than getting the relay state from the request, you should continue to get it from the message-context instead: `SAMLBindingSupport.getRelayState(messageContext);` The context should have the relay state as part of the first change in the controller, etc.

Please test and verify, and then follow up with integration tests.

Reply:
Hey @mmoayyed I just tried what you suggest and it doesn't seem to work. I've reverted the changes shown above, then extracted what the debug messages output in both places [here](https://github.com/linosgian/cas/blob/7ed196c377a3db9bb31ad1b62a42085ceec177ab/support/cas-server-support-saml-idp-web/src/main/java/org/apereo/cas/support/saml/web/idp/profile/AbstractSamlIdPProfileHandlerController.java#L463) and [here](https://github.com/linosgian/cas/blob/7ed196c377a3db9bb31ad1b62a42085ceec177ab/support/cas-server-support-saml-idp-web/src/main/java/org/apereo/cas/support/saml/web/idp/profile/builders/response/BaseSamlProfileSamlResponseBuilder.java#L106):

```
2021-07-19 12:03:44,634 DEBUG [org.apereo.cas.support.saml.web.idp.profile.AbstractSamlIdPProfileHandlerController] - <Using relay-state value [5QiOEdNhaACEl7SxzbSdwT1yvqTXM88Bjw1qi2FJ9h11oqX1b-xv8q8LPEFTdSHSCaZCF
5p4Fan6jo-2TJj8soMkp2OSarHO1wrgBVCyESY]>
```
And then, when the response is being built, we get the cached RelayState:
```
2021-07-19 12:03:44,923 TRACE [org.apereo.cas.support.saml.web.idp.profile.builders.response.BaseSamlProfileSamlResponseBuilder] - <RelayState is [PQHODZ1Tl-Dlas7mznDEnPvQIAmImzuKi7fFC3ZFevoqPaUWvfZaXGIL2vCoMZ1e_6
X6zeTzg-tLzJsxeQRoVEmYDepw76tRuBykZI65nd4]>
```

Reply:
As for testing, I was thinking of adding a test under `ci/puppeteer` to test the flow that's mentioned in the description. Did you have that kind of testing in mind?

Reply:
If you're up for it, that would be great.

Reply:
I've looked into the `ci/puppeteer` setup and it looks like `samltest.id` is used as an SP.
Unfortunately I couldn't find a way to pass `RelayState` so we verify that it gets sent back to the SP as we expect it.

Also, we'd need another SP to test the whole flow.

Shall we skip the test for the time being or do you have another idea?